docs/endpoint-recovery.md

```diff
@@ -0,0 +1,434 @@
+unresponsive, therefore it seems reasonable to assume it won't continue
+communicating with the endpoint unless the recovery succeeds. As such there's
+no action required when `.Connectivity` transitions to `Degraded`. However,
+if recovery succeeds, the transition of `.Connectivity` from `Degraded` to
+`Available` provides the signal to restart communicating with the endpoint.
+
+For applications in the second class it is likely the case that they
+haven't themselves invoked `.Recover` because they are yet to observe a
+communication failure with the endpoint. As such there's also no requirement
+for action when `.Connectivity` transitions to `Degraded` on behalf of another
+application. However, it may be the case that communication failures are
+subsequently observed. It's not necessary to invoke `.Recover` if
+`.Connectivity` is already in `Degraded`, though there should also be no
+significant consequences if it occurs. If no communication failures are observed
+while `.Connectivity` is in `Degraded` then there's also no action required if
+it transitions to `Available`.
+
+### `mctpd` behaviours over an endpoint recovery sequence
+
+For hotplug devices recovering from a reset `Get Endpoint ID` should yield an
+EID of 0.
+
+Note that while we're under the `Treclaim` period the device's allocated EID has
+not been quarantined to the reuse FIFO, and is also not a member of the dynamic
+allocation pool. For the purposes of allocation, the device is still assigned
+the EID it had prior to taking a reset. If a device at the physical address of
+the device in `Degraded` responds to `Get Endpoint ID` with an EID of 0 inside
+the `Treclaim` period and has a UUID that matches the device in `Degraded`, then
+it should be reassigned its allocated EID. At the end of this sequence the
+reassignment should not be observable to MCTP-application-layer beyond the
+endpoint's `.Connectivity` member transitioning from `Degraded` to `Available`.
+
+## MCTP device lifecycle and consuming events from `mctpd` D-Bus objects
+
+The general strategy for tracking endpoint lifecycles is [as follows]
+[dbus-spec-standard-interfaces-objectmanager]:
+
+[dbus-spec-standard-interfaces-objectmanager]: https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-objectmanager
+
+1. Add a match of the following form for event-driven updates to local
+   data-structures
+
+   ```
+   type='signal',sender='xyz.openbmc_project.MCTP',path_namespace='/xyz/openbmc_project/mctp'
+   ```
+
+2. Issue a call to `GetManagedObjects` `org.freedesktop.DBus.ObjectManager` on
+   `/xyz/openbmc_project/mctp` for initial population of local data-structures
+
+3. Issue
+
+### An example order of MCTP device events, consumed by `nvmesensor`
+
+1. A series of events yields a new FRU EEPROM device on an I2C bus already
+   registered as an MCTP endpoint.
+
+2. The FRU data is decoded and `SetupEndpoint` on the
```

Comment:
This is an example for the setting which BMC is BO or TMBO. How about in the PCIe interface when BMC can be endpoint?
In that case, instead of detect the FRU, BMC will receive the `prepare for discovery` signal from PCIe BO. As the previous discussion in `pcmi` discord, `mctpd` will response for that command and add the PCIe EID of BMC and also BO to mctp D-Bus interface.
Then `mctpd` will get the routing table from PCIe BO. Then how `mctpd` manages the EID in the received entry in the routing table? Does `mctpd` check the available status of device before adding it the MCTP D-Bus interface?